### PR TITLE
New functions for getting and updating Org vDC

### PR DESCRIPTION
**Pull request by vbauzysvmware**
Ref: terraform-providers/terraform-provider-vcd#237

Added functions for refreshing, getting and update Org vDC. Will be used in terraform.

### CODE REVIEW

File: govcd/org.go [hunk 1]
```diff
@@ -97,6 +97,84 @@ func (org *Org) GetVdcByName(vdcname string) (Vdc, error) {
 	return Vdc{}, nil
 }
 
+// Given an adminVdc with a valid HREF, the function refresh the adminVdc
+// and updates the adminVdc data. Otherwise if the function fails,
+// it returns an error.  Users should use refresh whenever they have
+// a stale vDC due to the creation/update/deletion of a resource
+// within the org or the vDC itself.
+func (adminVdc *AdminVdc) Refresh() error {
```
Comment (dataclouder) [inline govcd/org.go]: Suggestion:
```
// Given an adminVdc with a valid HREF, the function refresh the adminVdc
// and updates the adminVdc data. Returns an error on failure
// Users should use refresh whenever they suspect
// a stale vDC due to the creation/update/deletion of a resource
// within the the vDC itself.
```

File: govcd/org.go [hunk 2]
```diff
@@ -97,6 +97,84 @@ func (org *Org) GetVdcByName(vdcname string) (Vdc, error) {
...
+		"", "error refreshing vDC: %s", nil, unmarshalledAdminVdc)
+	if err != nil {
+		return err
+	}
+	adminVdc.AdminVdc = unmarshalledAdminVdc
+
+	return nil
+}
+
+// GetAdminVdcByName function uses user specifies valid vdc name and then returns a admin vDC object.
```
Comment (dataclouder) [inline govcd/org.go]:
```suggestion
// GetAdminVdcByName function uses a valid VDC name and returns a admin VDC object.
```

File: govcd/org.go [hunk 3]
```diff
@@ -97,6 +97,84 @@ func (org *Org) GetVdcByName(vdcname string) (Vdc, error) {
...
+			_, err := adminOrg.client.ExecuteRequest(vdcs.HREF, http.MethodGet,
+				"", "error getting vdc: %s", nil, adminVdc.AdminVdc)
+
+			return *adminVdc, err
+		}
+	}
+	return AdminVdc{}, nil
+}
+
+// Function UpdateAsync() updated vCD from current vCD struct contents.
```
Comment (dataclouder) [inline govcd/org.go]:
```suggestion
// UpdateAsync updates VDC from current VDC struct contents.
```

File: govcd/org.go [hunk 4]
```diff
@@ -97,6 +97,84 @@ func (org *Org) GetVdcByName(vdcname string) (Vdc, error) {
...
+func (adminVdc *AdminVdc) UpdateAsync() (Task, error) {
+
+	adminVdc.AdminVdc.Xmlns = types.XMLNamespaceVCloud
+
+	// Return the task
+	return adminVdc.client.ExecuteTaskRequest(adminVdc.AdminVdc.HREF, http.MethodPut,
+		types.MimeAdminVDC, "error updating vDC: %s", adminVdc.AdminVdc)
+}
+
+// Function Update() updated vCD from current vCD struct contents.
```
Comment (dataclouder) [inline govcd/org.go]:
```suggestion
// Update updates an Admin VDC from current VDC struct contents.
```

File: govcd/org_test.go [hunk 5]
```diff
@@ -420,3 +420,213 @@ func (vcd *TestVCD) Test_GetAdminCatalog(check *C) {
 		check.Assert(cat.AdminCatalog.Description, Equals, vcd.config.VCD.Catalog.Description)
 	}
 }
+
+// Tests Refresh for Vdc by updating the Vdc and then asserting if the
```
Comment (dataclouder) [inline govcd/org_test.go]:
```suggestion
// Tests Refresh for VDC by updating it and then asserting if the
```

File: govcd/org_test.go [hunk 6]
```diff
@@ -420,3 +420,213 @@ func (vcd *TestVCD) Test_GetAdminCatalog(check *C) {
...
+		err = vdc.DeleteWait(true, true)
+		check.Assert(err, IsNil)
+	}
+	err = adminOrg.CreateVdcWait(vdcConfiguration)
+	check.Assert(err, IsNil)
+	AddToCleanupList(vdcConfiguration.Name, "vdc", vcd.org.Org.Name, check.TestName())
+	return adminOrg, err, vdcConfiguration
+}
+
+// Tests Vdc by updating the Vdc and then asserting if the
```
Comment (dataclouder) [inline govcd/org_test.go]:
```suggestion
// Tests VDC by updating it and then asserting if the
```

File: types/v56/types.go [hunk 7]
```diff
@@ -2435,3 +2437,27 @@ type PortGroupRecordType struct {
 	ScopeType     int     `xml:"scopeType,attr,omitempty"` // Scope of network using the portgroup(1=Global, 2=Organization, 3=vApp)
 	Link          []*Link `xml:"Link,omitempty"`
 }
+
+// Represents org VDC Network
+// Reference: vCloud API 27.0 - Org VDC Network
+// https://code.vmware.com/apis/72/doc/doc/types/QueryResultOrgVdcNetworkRecordType.html
+type QueryResultOrgVdcNetworkRecordType struct {
```
Comment (lvirbalas) [inline types/v56/types.go]: Just curious, how come this struct was not needed before and is needed now?

File: types/v56/types.go [hunk 8]
```diff
@@ -394,8 +394,8 @@ type AdminVdc struct {
 	Xmlns string `xml:"xmlns,attr"`
 	Vdc
 
-	ResourceGuaranteedMemory float64    `xml:"ResourceGuaranteedMemory,omitempty"`
-	ResourceGuaranteedCpu    float64    `xml:"ResourceGuaranteedCpu,omitempty"`
+	ResourceGuaranteedMemory string     `xml:"ResourceGuaranteedMemory,omitempty"`
```
Comment (Didainius) [inline types/v56/types.go]: Any reason why these become strings?

Comment (vbauzysvmware) [reply types/v56/types.go]: Yes, I need to serialize `0` and not ignore them.

Comment (lvirbalas) [reply types/v56/types.go]: Won't this cause issue like we had with https://github.com/vmware/go-vcloud-director/pull/198 ?

Comment (vbauzysvmware) [reply types/v56/types.go]: In this case no, it just number for how much percentage is guarantee. And don't care - I didn't find solution how to serialize int64=0 -> it's ignored by default, as string="" and similar values

Comment (Didainius) [reply types/v56/types.go]: What if you skip the `omitempty`?

Comment (vbauzysvmware) [reply types/v56/types.go]: I can't send no value, but should

Comment (vbauzysvmware) [reply types/v56/types.go]: With Dainius we find a little better approach

Comment (lvirbalas) [reply types/v56/types.go]: It's pretty cool!

File: govcd/org.go (hunk 1 above)
Comment (vbauzysvmware) [reply govcd/org.go]: Tx. Updated

File: govcd/org.go (hunk 2 above)
Comment (vbauzysvmware) [reply govcd/org.go]: Tx. Updated

File: govcd/org.go (hunk 3 above)
Comment (vbauzysvmware) [reply govcd/org.go]: Tx. Updated

File: govcd/org.go (hunk 4 above)
Comment (vbauzysvmware) [reply govcd/org.go]: Tx. Updated

File: govcd/org_test.go (hunk 6 above)
Comment (vbauzysvmware) [reply govcd/org_test.go]: Tx. Updated

File: govcd/org_test.go (hunk 5 above)
Comment (vbauzysvmware) [reply govcd/org_test.go]: Tx. Updated

File: types/v56/types.go [hunk 9]
```diff
@@ -367,22 +367,22 @@ type Vdc struct {
 	ID           string `xml:"id,attr,omitempty"`
 	OperationKey string `xml:"operationKey,attr,omitempty"`
 	Name         string `xml:"name,attr"`
-	Status       int    `xml:"status,attr,omitempty"`
+	Status       string `xml:"status,attr,omitempty"`
```
[the remark this answers is not in the document]
Comment (vbauzysvmware) [reply types/v56/types.go]: Changed

File: govcd/org.go [hunk 10]
```diff
@@ -97,6 +97,84 @@ func (org *Org) GetVdcByName(vdcname string) (Vdc, error) {
 	return Vdc{}, nil
 }
 
+// Given an adminVdc with a valid HREF, the function refresh the adminVdc
+// and updates the adminVdc data. Otherwise if the function fails,
+// it returns an error.  Users should use refresh whenever they have
+// a stale vDC due to the creation/update/deletion of a resource
+// within the org or the vDC itself.
+func (adminVdc *AdminVdc) Refresh() error {
+	if *adminVdc == (AdminVdc{}) {
```
[the remark this answers is not in the document]
Comment (vbauzysvmware) [reply govcd/org.go]: added

File: types/v56/types.go (hunk 7 above)
Comment (vbauzysvmware) [reply types/v56/types.go]: Left as was doing other POC. This just for future use

File: govcd/org_test.go [hunk 11]
```diff
@@ -420,3 +420,213 @@ func (vcd *TestVCD) Test_GetAdminCatalog(check *C) {
...
+	check.Assert(updatedVdc.AdminVdc.Description, Equals, updateDescription)
+	check.Assert(updatedVdc.AdminVdc.ComputeCapacity[0].CPU.Allocated, Equals, computeCapacity[0].CPU.Allocated)
+	check.Assert(updatedVdc.AdminVdc.IsEnabled, Equals, false)
+	check.Assert(updatedVdc.AdminVdc.IsThinProvision, Equals, false)
+	check.Assert(updatedVdc.AdminVdc.NetworkQuota, Equals, quota)
+	check.Assert(updatedVdc.AdminVdc.VMQuota, Equals, quota)
+	check.Assert(updatedVdc.AdminVdc.OverCommitAllowed, Equals, false)
+	check.Assert(updatedVdc.AdminVdc.VCpuInMhz, Equals, vCpu)
+	check.Assert(updatedVdc.AdminVdc.UsesFastProvisioning, Equals, false)
+	check.Assert(updatedVdc.AdminVdc.ResourceGuaranteedCpu, Equals, guaranteed)
```
Comment (Didainius) [inline govcd/org_test.go]: Probably needs de-referencing as test fails:

```
START: org_test.go:548: TestVCD.Test_UpdateVdc
org_test.go:605:
    check.Assert(updatedVdc.AdminVdc.ResourceGuaranteedCpu, Equals, guaranteed)
... obtained *float64 = (*float64)(0xc000026150)
... expected float64 = 0.6
```

Comment (vbauzysvmware) [reply govcd/org_test.go]: fixed

File: govcd/org_test.go [hunk 12]
```diff
@@ -420,3 +421,213 @@ func (vcd *TestVCD) Test_GetAdminCatalog(check *C) {
...
+	check.Assert(err, IsNil)
+	AddToCleanupList(TestRefreshOrgVdc, "vdc", vcd.org.Org.Name, check.TestName())
+
+	// Test Refresh on vdc
+	err = adminVdc.Refresh()
+	check.Assert(err, IsNil)
+	check.Assert(adminVdc.AdminVdc.Name, Equals, TestRefreshOrgVdc)
+}
+
+func setupVDc(vcd *TestVCD, check *C) (AdminOrg, error, *types.VdcConfiguration) {
```
[the remark this answers is not in the document]
Comment (vbauzysvmware) [reply govcd/org_test.go]: Fixed

**Pull request merged**